include/boost/geometry/algorithms/detail/overlay/discard_duplicate_turns.hpp

```diff
@@ -50,47 +55,62 @@ inline bool corresponding_turn(Turn const& turn, Turn const& start_turn,
                                Geometry0 const& geometry0,
                                Geometry1 const& geometry1)
 {
+    std::size_t count = 0;
     for (std::size_t i = 0; i < 2; i++)
     {
         for (std::size_t j = 0; j < 2; j++)
         {
-            if (turn.operations[i].seg_id == start_turn.operations[j].seg_id
-                && (turn.method == method_touch
-                    || turn.method == method_touch_interior))
+            if (same_multi_and_ring_id(turn.operations[i].seg_id,
+                                       start_turn.operations[j].seg_id))
             {
-                // Verify if the other operation is consecutive
-                return is_consecutive(turn.operations[1 - i].seg_id,
+                // Verify if all operations are consecutive
+                if (is_consecutive(turn.operations[i].seg_id,
+                                   start_turn.operations[j].seg_id,
+                                   geometry0, geometry1)
+                    && is_consecutive(turn.operations[1 - i].seg_id,
                                       start_turn.operations[1 - j].seg_id,
-                                      geometry0, geometry1);
+                                      geometry0, geometry1))
+
+                {
+                    count++;
+                }
             }
         }
     }
-    return false;
+    // An intersection is located on exactly two rings
+    // The corresponding turn, if any, should be located on the same two rings.
+    return count == 2;
 }
 
-
 template <typename Turns, typename Geometry0, typename Geometry1>
 inline void discard_duplicate_start_turns(Turns& turns,
                                           Geometry0 const& geometry0,
                                           Geometry1 const& geometry1)
 {
-    // Start turns are generated, in case the previous turn is missed
-    // But often it is not missed, and then it should still be deleted
+    // Start turns are generated, in case the previous turn is missed.
+    // But often it is not missed, and then it should be deleted.
     // This is how it can be
     // (in float, collinear, points far apart due to floating point precision)
     //   [m, i s:0, v:6 1/1 (1) // u s:1, v:5  pnt (2.54044, 3.12623)]
     //   [s, i s:0, v:7 0/1 (0) // u s:1, v:5  pnt (2.70711, 3.29289)]
 
-    // 1 Build map (segment->turnids) of start turns
-    std::map<segment_identifier, std::vector<std::size_t>> start_turns_per_segment;
+    using multi_and_ring_id_type = std::pair<signed_size_type, signed_size_type>;
+
+    auto adapted_id_of = [](segment_identifier const& seg_id)
```

Comment:
This is a function so to should probably be called `adapt_id_of` or `to_multi_and_ring_id` ot something like that. But I'm ok with the current name if you don't feel to change it.

Reply:
Make sense. I will change it and then merge, thanks.